File: mcp/vault/scripts/vault_hook.py
import json
import os
import re
import sys
from pathlib import Path

from vault_core import (
    DEFAULT_HOME,
    add_journal_entry,
    add_record,
    ensure_active_vault,
    vault_status,
)


HOME = str(Path(os.environ.get("VAULT_HOME", DEFAULT_HOME)).expanduser().resolve())


def read_stdin():
    raw = sys.stdin.read().strip()
    if not raw:
        return {}
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return {}


def session_id(input_data):
    value = str(
        input_data.get("session_id")
        or input_data.get("sessionId")
        or input_data.get("session")
        or "unknown"
    )
    return re.sub(r"[^a-zA-Z0-9_-]", "", value)


def transcript_to_text(transcript_path):
    if not transcript_path:
        return ""
    file_path = Path(transcript_path).expanduser().resolve()
    if not file_path.exists():
        return ""
    raw = file_path.read_text(encoding="utf-8")
    if file_path.suffix.lower() in {".jsonl", ".json"}:
        snippets = []
        for line in raw.splitlines():
            if not line.strip():
                continue
            try:
                entry = json.loads(line)
            except json.JSONDecodeError:
                continue
            text = (
                entry.get("message", {}).get("content")
                or entry.get("payload", {}).get("message")
                or entry.get("text")
            )
            if isinstance(text, str) and text.strip():
                snippets.append(text.strip())
        return "\n".join(snippets[-40:])
    return raw


def emit(payload):
    sys.stdout.write(json.dumps(payload) + "\n")


def run():
    command = sys.argv[1] if len(sys.argv) > 1 else "session-start"
    input_data = read_stdin()
    active = ensure_active_vault(HOME)

    if not active["active"]:
        emit(vault_status(HOME))
        return

    sid = session_id(input_data)
    transcript_path = input_data.get("transcript_path") or input_data.get("transcriptPath") or ""
    transcript = transcript_to_text(transcript_path)
    title = input_data.get("title") or input_data.get("session_title") or f"{command} checkpoint"
    note = input_data.get("note") or f"{command} for {sid}"

    if transcript:
        add_record(
            vault_path=active["active"]["path"],
            title=title,
            content=transcript,
            tags=["hook", command],
            source=transcript_path,
            kind="hook-transcript",
            metadata={"sessionId": sid, "hook": command},
        )

    add_journal_entry(
        vault_path=active["active"]["path"],
        session_id=sid,
        entry_type=command,
        note=note,
        metadata={"transcriptPath": transcript_path},
    )

    emit(
        {
            "ok": True,
            "command": command,
            "sessionId": sid,
            "wroteTranscript": bool(transcript),
            "vault": active["active"]["name"],
        }
    )


if __name__ == "__main__":
    try:
        run()
    except Exception as error:
        emit({"ok": False, "error": str(error)})
        sys.exit(1)

File: skills/superdeveloper/mcp/vault/scripts/vault-hook.js
import fs from 'fs/promises';
import path from 'path';
import os from 'os';
import {
  addJournalEntry,
  addRecord,
  DEFAULT_HOME,
  ensureActiveVault,
  vaultStatus,
} from './vault-core.js';

const HOME = process.env.VAULT_HOME ? path.resolve(process.env.VAULT_HOME) : DEFAULT_HOME;

const readStdin = async () => {
  const chunks = [];
  for await (const chunk of process.stdin) {
    chunks.push(Buffer.from(chunk));
  }
  const raw = Buffer.concat(chunks).toString('utf8').trim();
  if (!raw) return {};
  try {
    return JSON.parse(raw);
  } catch {
    return {};
  }
};

const sessionId = (input) =>
  String(input.session_id || input.sessionId || input.session || 'unknown').replace(/[^a-zA-Z0-9_-]/g, '');

const transcriptToText = async (transcriptPath) => {
  if (!transcriptPath) return '';
  const resolved = path.resolve(transcriptPath);
  const ext = path.extname(resolved).toLowerCase();
  try {
    const raw = await fs.readFile(resolved, 'utf8');
    if (ext === '.jsonl' || ext === '.json') {
      const lines = raw.split('\n').filter(Boolean);
      const snippets = [];
      for (const line of lines) {
        try {
          const entry = JSON.parse(line);
          const text = entry?.message?.content || entry?.payload?.message || entry?.text;
          if (typeof text === 'string' && text.trim()) {
            snippets.push(text.trim());
          }
        } catch {
          // ignore
        }
      }
      return snippets.slice(-40).join('\n');
    }
    return raw;
  } catch {
    return '';
  }
};

const emit = async (payload) => {
  process.stdout.write(`${JSON.stringify(payload)}\n`);
};

const run = async () => {
  const command = process.argv[2] || 'session-start';
  const input = await readStdin();
  const active = await ensureActiveVault(HOME);

  if (!active.active) {
    await emit(await vaultStatus(HOME));
    return;
  }

  const sid = sessionId(input);
  const transcript = await transcriptToText(input.transcript_path || input.transcriptPath || '');
  const title = input.title || input.session_title || `${command} checkpoint`;
  const note = input.note || `${command} for ${sid}`;

  if (transcript) {
    await addRecord({
      vaultPath: active.active.path,
      title,
      content: transcript,
      tags: ['hook', command],
      source: input.transcript_path || input.transcriptPath || '',
      kind: 'hook-transcript',
      metadata: { sessionId: sid, hook: command },
    });
  }

  await addJournalEntry({
    vaultPath: active.active.path,
    sessionId: sid,
    entryType: command,
    note,
    metadata: {
      transcriptPath: input.transcript_path || input.transcriptPath || '',
    },
  });

  await emit({
    ok: true,
    command,
    sessionId: sid,
    wroteTranscript: Boolean(transcript),
    vault: active.active.name,
  });
};

run().catch(async (error) => {
  await emit({ ok: false, error: error instanceof Error ? error.message : String(error) });
  process.exit(1);
});
